### PR TITLE
nll: correctly deal with bivariance

### LINKED ISSUE #104409: mir typeck: subtyping cannot deal with bivariance
```rust
trait Trait {
    type Assoc;
}

struct Foo<T, U>(T)
where
    T: Trait<Assoc = U>;

impl Trait for for<'a> fn(&'a ()) {
    type Assoc = u32;
}
impl Trait for fn(&'static ()) {
    type Assoc = String;
}
    
fn foo(x: Foo<for<'a> fn(&'a ()), u32>) -> Foo<fn(&'static ()), String> {
    x
}

fn main() {}
```
results in the following ICE:
```rust
error: internal compiler error: broken MIR in DefId(0:15 ~ example[9884]::foo) (_0 = move _1): bad assignment (Foo<fn(&()), std::string::String> = Foo<for<'a> fn(&'a ()), u32>): NoSolution
  --> <source>:17:5
   |
17 |     x
   |     ^

thread 'rustc' panicked at 'aborting due to `-Z treat-err-as-bug=1`', compiler/rustc_errors/src/lib.rs:1561:30
stack backtrace:
----------- SNIP ----------------------
  17:     0x7fe0e8c751b3 - <rustc_errors[1020869c0a53ac34]::HandlerInner>::panic_if_treat_err_as_bug
  18:     0x7fe0e8c73258 - <rustc_errors[1020869c0a53ac34]::HandlerInner>::emit_diagnostic
  19:     0x7fe0ea774481 - <rustc_errors[1020869c0a53ac34]::HandlerInner>::emit_diag_at_span::<rustc_span[efd8b0f1895919e0]::span_encoding::Span>
  20:     0x7fe0ea774603 - <rustc_errors[1020869c0a53ac34]::HandlerInner>::span_bug::<rustc_span[efd8b0f1895919e0]::span_encoding::Span, &str>
  21:     0x7fe0ea773c55 - <rustc_errors[1020869c0a53ac34]::Handler>::delay_span_bug::<rustc_span[efd8b0f1895919e0]::span_encoding::Span, &str>
  22:     0x7fe0e9228157 - <rustc_borrowck[e2ad9be94bf77b7a]::type_check::TypeChecker>::typeck_mir
  23:     0x7fe0e919c351 - rustc_borrowck[e2ad9be94bf77b7a]::type_check::type_check
  24:     0x7fe0e9187179 - rustc_borrowck[e2ad9be94bf77b7a]::nll::compute_regions
  25:     0x7fe0e915c6e8 - rustc_borrowck[e2ad9be94bf77b7a]::do_mir_borrowck
  26:     0x7fe0e9b8f296 - rustc_borrowck[e2ad9be94bf77b7a]::mir_borrowck
```
[godbolt](https://godbolt.org/z/WdM3qPM63)

This should compile

### PR DESCRIPTION
**Pull request by lcnr**
fixes #104409

when in a bivariant context, relating stuff should always trivially succeed. Also changes the mir validator to correctly deal with higher ranked regions.

r? types cc @RalfJung

### CODE REVIEW

File: compiler/rustc_const_eval/src/transform/validate.rs [hunk 1]
```diff
@@ -87,27 +86,40 @@ pub fn equal_up_to_regions<'tcx>(
...
-                ty_op: |ty| ty,
-            },
-        )
-    };
-    tcx.infer_ctxt().build().can_eq(param_env, normalize(src), normalize(dest)).is_ok()
+    // Check for subtyping in either direction.
+    is_subtype(tcx, param_env, src, dest) || is_subtype(tcx, param_env, dest, src)
 }
 
+pub fn is_subtype<'tcx>(
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/transform/validate.rs:93]: The doc comment should say which one is the sub- and which the supertype.

**Comment by RalfJung** [comment]
On a conceptual level this makes sense to me, but I understand absolutely nothing of what goes on inside `is_subtype`^^

**Comment by lcnr** [comment]
woops, didn't actually include the changes to the `nll_relate::TypeRelating` :sweat_smile: no wonder the CI failed.

**Pull request merged**